text/differential-layer-format.md

```diff
@@ -0,0 +1,143 @@
+When in memory, with the 8-byte LSN, we have 2 bytes left to store some more
+metadata of the branch in the BranchPtr, such as type info and other relevant
+information.
+
+
+The per-page data layout of the Differential layer will thus be:
+
+```text
+  Page data:
+    ... metadata
+    page_versions: sorted array of (Lsn -> BranchPtr) pairs, in descending order of LSN
+      Lsn -> BranchPtr:
+        Lsn: 8 bytes
+        BranchPtr: 8 bytes
+          offset_to_branch: 6 bytes
+          branch type specific info: 2 bytes
+          ...
+      Lsn -> BranchPtr -> ...
+      Lsn -> BranchPtr -> ...
+      Lsn -> BranchPtr -> ...
+      ...
+```
+
+Note that the definition of BranchPtr is partially dependent on what data we
+want to store in the BranchPtr, and what data we want to store locally in the
+Branch. Either way, the overhead of one branch in the page versions array is
+16 bytes.
+
+
+
+### Page index
+The Delta layer does not use an index to check if a page exists in the Layer, but
+instead uses the combined index for on (page, pageversion). This means that finding
+a page in a Delta layer is log(n) (for n = number of page versions in the Layer).
+This is very expensive and not really sustainable when we're looking forward:
+larger segments means larger amounts of page versions in a layer, which means
+larger lookup times.  
+In the Delta layer there is also no description on when a page was last modified,
+so any failed lookup in the Delta layer must search back to the previous layer,
+making such lookups in the worst case O(log N * M) for N pageversions per layer
+and M layers. This currently is not too much of an issue, because this M is
+limited to 1 due to how our layers work: we always have an image layer
+immediately before the delta layer, which limits the number of Delta layers we
+access to 1.
+
+To limit the write amplification caused by the need for Image layers, the
+Differential layer uses a mapping (implemented as bitmap for O(log1024(n)) access,
+but could be any int->data map) to store the page index. This page index stores, for all pages
+modified since the last snapshot, either a pointer to the page data, or the LSN
+of the last time the page was modified after the last snapshot. This means that
+using one lookup we can discern whether we can access the page data, and if not,
+where we should look instead.
+
+These improvements mean that the search for some old page version goes from 
+`O(m log N)` to `O(log n)`; a nice improvement.
+
+### Segment size
+The segment size of the current layer formats is very small; which means that
+the number of layers is relatively high. As each layer has some constant
+overhead, it should be beneficial to increase layer size to something in the
+order of gigabytes, instead of megabytes.
+
+The Differential layer format is designed for segment sizes of up to 8GiB,
+almost 1000 times larger than the current segment size. 
+
```

Comment:
Write amplification for image layers is proportional to layer size. Certainly we can postpone generation of image layers. But not so long. And worst case scenarios may cause x1000 and even more write amplification for 1Gb segments.
Certainly we can use incremental image layers. But in this case get_page_at_lsn will have to read many files and perform searches in them.

Reply:
That is exactly why I implement a page index in each layer, which either points to the branches for that page in this layer, or contains the last LSN that this page was modified. No need for intermediate image layers; only one base image layer required (to allow us to remove older WAL). Updates can then be flushed down onto that one base image layer when the layer containing those updates are removed from the local disk.